listings/applications/advanced_factory/src/campaign.cairo

```diff
@@ -0,0 +1,286 @@
+        status: Status
+    }
+
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        #[flat]
+        OwnableEvent: ownable_component::Event,
+        ContributableEvent: contributable_component::Event,
+        ContributionMade: ContributionMade,
+        Claimed: Claimed,
+        Closed: Closed,
+        Upgraded: Upgraded,
+        Withdrawn: Withdrawn,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct ContributionMade {
+        #[key]
+        pub contributor: ContractAddress,
+        pub amount: u256,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct Claimed {
+        pub amount: u256,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct Closed {
+        pub reason: ByteArray,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct Upgraded {
+        pub implementation: ClassHash
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct Withdrawn {
+        #[key]
+        pub contributor: ContractAddress,
+        pub amount: u256,
+    }
+
+    pub mod Errors {
+        pub const NOT_FACTORY: felt252 = 'Caller not factory';
+        pub const ENDED: felt252 = 'Campaign already ended';
+        pub const STILL_ACTIVE: felt252 = 'Campaign not ended';
+        pub const ZERO_DONATION: felt252 = 'Donation must be > 0';
+        pub const ZERO_TARGET: felt252 = 'Target must be > 0';
+        pub const ZERO_DURATION: felt252 = 'Duration must be > 0';
+        pub const ZERO_FUNDS: felt252 = 'No funds to claim';
+        pub const TRANSFER_FAILED: felt252 = 'Transfer failed';
+        pub const TITLE_EMPTY: felt252 = 'Title empty';
+        pub const CLASS_HASH_ZERO: felt252 = 'Class hash cannot be zero';
+        pub const FACTORY_ZERO: felt252 = 'Factory address cannot be zero';
+        pub const CREATOR_ZERO: felt252 = 'Creator address cannot be zero';
+        pub const TARGET_NOT_REACHED: felt252 = 'Target not reached';
+        pub const TARGET_ALREADY_REACHED: felt252 = 'Target already reached';
+        pub const NOTHING_TO_WITHDRAW: felt252 = 'Nothing to withdraw';
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        owner: ContractAddress,
+        title: ByteArray,
+        description: ByteArray,
+        target: u256,
+        duration: u64,
+        token_address: ContractAddress
+    ) {
+        assert(owner.is_non_zero(), Errors::CREATOR_ZERO);
+        assert(title.len() > 0, Errors::TITLE_EMPTY);
+        assert(target > 0, Errors::ZERO_TARGET);
+        assert(duration > 0, Errors::ZERO_DURATION);
+
+        self.token.write(IERC20Dispatcher { contract_address: token_address });
+
+        self.title.write(title);
+        self.target.write(target);
+        self.description.write(description);
+        self.end_time.write(get_block_timestamp() + duration);
+        self.factory.write(get_caller_address());
+        self.ownable._init(owner);
+        self.status.write(Status::ACTIVE)
```

Comment:
suggestion: add status `PENDING` with a owner guarded function `start` to change the status from PENDING to ACTIVE at anytime